File: internal/rule/loginform.go
package rule

import (
	"regexp"
	"strings"
)

var passwordInputRe = regexp.MustCompile(`(?i)<input[^>]+type=["']password["'][^>]*>`)
var emailInputRe = regexp.MustCompile(`(?i)<input[^>]+type=["'](?:email|text)["'][^>]*name=["'](?:email|user|login|username)["'][^>]*>`)
var formActionRe = regexp.MustCompile(`(?i)<form[^>]+action=["']([^"']+)["'][^>]*>`)

// LoginFormDetector flags pages that contain credential harvesting forms.
type LoginFormDetector struct{}

func (l *LoginFormDetector) Name() string { return "login-form" }

func (l *LoginFormDetector) Evaluate(ev Evidence) ([]Label, error) {
	html := strings.ToLower(ev.HTML)
	var labels []Label

	hasPassword := passwordInputRe.MatchString(html)
	hasEmail := emailInputRe.MatchString(html)

	if !hasPassword {
		return nil, nil
	}

	confidence := 0.5
	detail := "password input detected"
	if hasEmail {
		confidence = 0.8
		detail = "login form with email/username + password inputs"
	}

	if m := formActionRe.FindStringSubmatch(ev.HTML); len(m) > 1 {
		action := m[1]
		if action != "" && !strings.Contains(action, ev.Domain) {
			confidence = 0.95
			detail += "; form posts to external domain: " + action
		}
	}

	labels = append(labels, Label{
		Rule:       l.Name(),
		Name:       "login-form",
		Confidence: confidence,
		Detail:     detail,
	})

	return labels, nil
}

File: internal/project/templates/rules/login-form.ts
import { defineRule } from "@openbpl/sdk";

export default defineRule({
  name: "login-form",
  description: "Detects credential harvesting forms with password inputs",

  evaluate({ evidence, brand }) {
    const html = evidence.html.toLowerCase();

    const hasPassword = /<input[^>]+type=["']password["'][^>]*>/i.test(html);
    if (!hasPassword) return null;

    const hasEmail =
      /<input[^>]+type=["'](?:email|text)["'][^>]*name=["'](?:email|user|login|username)["'][^>]*>/i.test(
        html
      );

    let confidence = 0.5;
    let detail = "password input detected";

    if (hasEmail) {
      confidence = 0.8;
      detail = "login form with email/username + password inputs";
    }

    // Check if form posts to external domain
    const formAction = /<form[^>]+action=["']([^"']+)["'][^>]*>/i.exec(
      evidence.html
    );
    if (formAction && formAction[1]) {
      const action = formAction[1];
      if (action && !action.includes(evidence.domain)) {
        confidence = 0.95;
        detail += `; form posts to external domain: ${action}`;
      }
    }

    return { name: "login-form", confidence, detail };
  },
});
